src/lib/coreaudio/CoreAudioDevice.hxx

```diff
@@ -0,0 +1,119 @@
+#include <CoreAudio/CoreAudio.h>
+
+class CoreAudioDevice
+{
+public:
+	CoreAudioDevice();
+	explicit CoreAudioDevice(AudioDeviceID dev_id);
+	virtual ~CoreAudioDevice();
+	// Initialize and search for device name, falls back to system default device
+	bool Open(const std::string &device_name);
+	// Restore settings and close the device
+	void Close();
+	// Starts the device (i.e. enables CoreAudio HAL to callback for data)
+	void Start();
+	// Stops the device (used for pause, no data will be requested)
+	void Stop();
+	AudioDeviceID GetId() {return device_id;}
+	std::string GetName() const;
+	bool IsPlanar() {return is_planar;}
+	UInt32 GetTotalOutputChannels() const;
+	UInt32 GetNumChannelsOfStream(UInt32 stream_idx) const;
+	UInt32 GetStreamIdx() {return output_stream_idx;}
+	bool GetStreams(AudioStreamIdList *stream_id_list);
+	// Turn on exclusive device usage
+	bool SetHogStatus(bool hog);
+	pid_t GetHogStatus();
+	bool HasVolume() {return has_volume;}
+	bool SetCurrentVolume(Float32 vol);
+	Float32 GetCurrentVolume();
+	UInt32 GetBufferSize();
+	bool SetBufferSize(UInt32 size);
+	bool AddIOProc(AudioDeviceIOProc io_proc, void* callback_data);
+	bool RemoveIOProc();
+	/** Main method called for every MPD format change.
+	 *	Uses scoring to find the best matching output
+	 *	format for MPD AudioFormat format. In case
+	 *	prefer_unmixable is set, unmixable integer
+	 *	formats are used.
+	 */
+	bool SetFormat(const AudioFormat &format, bool prefer_unmixable);
+	// Return physical format used internally by the device output stream
+	AudioStreamBasicDescription GetPhysFormat();
+	// Return the IO format that the device expects to be fed
+	AudioStreamBasicDescription GetIOFormat();
+	
+private:
+
+	bool started;
+	// To identify devices where streams are single channels
+	bool is_planar;
+	AudioDeviceID device_id;
+	UInt32 output_stream_idx;
+	AudioStreamBasicDescription output_format;
+	AudioDeviceIOProc io_proc;
+	CoreAudioStream output_stream;
+	bool has_volume;
+	pid_t hog_pid;
+	UInt32 buffer_size_restore;
+	
+	// Collect information from the device
+	bool Enumerate();
+	
+	
+	/** Scores a format based on:
+	 *	1. Matching sample rate (or integer fraction / multiple)
+	 *	2. Matching bits per channel (or higher).
+	 *	3. Matching number of channels (or higher).
+	 *
+	 *	The ASBD which should be scored against the
+	 *	MPD AudioFormat format which should be matched
+	 *	to be passed.
+	 */
+	float ScoreFormat(const AudioStreamBasicDescription &format_desc, const AudioFormat &format) const;
+	
+	/** Scores a samplerate based on:
+	 * 	1. Prefer exact match
+	 *	2. Prefer exact multiple of source samplerate
+	 *	The destination samplerate to score is compared against
+	 *	the requested source rate of the audio format.
+	 */
+	float ScoreSampleRate(Float64 destination_rate, unsigned int source_rate) const;
+	
+	typedef struct
+	{
+		AudioStreamID stream_id;
+		StreamFormatList format_list;
+		UInt32 num_channels;
+	} stream_info;
```

Comment:
Coding style violation: CamelCase for struct names.

Reply:
adjusted